library/dnet.c

```diff
@@ -393,6 +406,375 @@ static int dnet_iterator_callback_send(void *priv, void *data, uint64_t dsize)
+					(unsigned long long)re->user_flags,
+					(unsigned long long)re->timestamp.tsec, (unsigned long long)re->timestamp.tnsec,
+					re->status, (unsigned long long)re->size,
+					(unsigned long long)re->iterated_keys, (unsigned long long)re->total_keys,
+					send->write_error);
+
+			atomic_dec(&send->writes_pending);
+			pthread_cond_broadcast(&send->write_wait);
+
+			dnet_server_send_put(send);
+			free(wp);
+			return err;
+		}
+	}
+
+	return err;
+}
+
+struct dnet_server_send_ctl *dnet_server_send_alloc(void *state, struct dnet_cmd *cmd, uint64_t iflags,
+		int *groups, int group_num)
+{
+	int err;
+	struct dnet_net_state *st = state;
+	struct dnet_server_send_ctl *ctl;
+
+	dnet_state_get(st);
+
+	ctl = malloc(sizeof(struct dnet_server_send_ctl) + sizeof(int) * group_num);
+	if (!ctl) {
+		err = -ENOMEM;
+		goto err_out_exit;
+	}
+
+	memset(ctl, 0, sizeof(struct dnet_server_send_ctl));
+
+	ctl->state = state;
+	ctl->cmd = *cmd;
+	ctl->iflags = iflags;
+	ctl->groups = (int *)(ctl + 1);
+	memcpy(ctl->groups, groups, sizeof(int) * group_num);
+	ctl->group_num = group_num;
+	atomic_set(&ctl->writes_pending, 0);
+	atomic_set(&ctl->refcnt, 1);
+
+	err = pthread_cond_init(&ctl->write_wait, NULL);
+	if (err) {
+		err = -err;
+		dnet_log(st->n, DNET_LOG_ERROR, "Failed to initialize server send condition variable: %d", err);
+		goto err_out_free;
+	}
+
+	err = pthread_mutex_init(&ctl->write_lock, NULL);
+	if (err) {
+		err = -err;
+		dnet_log(st->n, DNET_LOG_ERROR, "Failed to initialize server send write lock: %d", err);
+		goto err_out_cond_destroy;
+	}
+
+	return ctl;
+
+err_out_cond_destroy:
+	pthread_cond_destroy(&ctl->write_wait);
+err_out_free:
+	free(ctl);
+err_out_exit:
+	dnet_state_put(st);
+	return NULL;
+}
+
+static int dnet_server_send_cleanup(struct dnet_server_send_ctl *ctl)
+{
+	int err = ctl->write_error;
+
+	/*
+	 * This bit will not be set for iterator request.
+	 * It can be set for async operations,
+	 * for example DNET_CMD_SEND handler sets up a bunch of async
+	 * WRITE requests and returns. When all writes will have been completed,
+	 * this cleanup function is invoked, but since they were async,
+	 * client hasn't received ACK and there is no blocking
+	 * thread waiting for all requests to complete (like in iterator command)
+	 */
+	if (ctl->cmd.flags & DNET_FLAGS_NEED_ACK) {
+		dnet_send_ack(ctl->state, &ctl->cmd, err, 0);
+	}
+
+	pthread_cond_destroy(&ctl->write_wait);
+	pthread_mutex_destroy(&ctl->write_lock);
+
+	dnet_state_put(ctl->state);
+	free(ctl);
+
+	return err;
+}
+
+struct dnet_server_send_ctl *dnet_server_send_get(struct dnet_server_send_ctl *ctl)
+{
+	atomic_inc(&ctl->refcnt);
+	return ctl;
+}
+int dnet_server_send_put(struct dnet_server_send_ctl *ctl)
+{
+	if (atomic_dec_and_test(&ctl->refcnt))
+		return dnet_server_send_cleanup(ctl);
+
+	return 0;
+}
+
+static int dnet_server_send_sync(struct dnet_server_send_ctl *ctl)
+{
+	while (atomic_read(&ctl->writes_pending) > 0) {
```

Comment:
When someone sets `st->__need_exit` flag appropriate network or IO thread starts killing the state and removes all pending transactions, which in turn invoke server-send completion callbacks which decrease this variable.

There is intermediate sync which waits until 500 writes out of 1000 pending operations complete, it checks `st->__need_exit` flag though, since it can be interrupted in the middle of the iteration, while `dnet_server_send_sync()` is only called to wait for the last 1000 writes to complete.